src/Lookup.hs

```diff
@@ -30,90 +28,56 @@ lookupInEnv path@(SymPath (p : ps) name) env =
-findAllGlobalVariables env =
-  concatMap finder (envBindings env)
-  where
-    finder :: Binder -> [Binder]
-    finder def@(Binder _ (XObj (Lst (XObj Def _ _ : _)) _ _)) =
-      [def]
-    finder (Binder _ (XObj (Mod innerEnv) _ _)) =
-      findAllGlobalVariables innerEnv
-    finder _ =
-      []
-
--- | Find all the possible (imported) symbols that could be referred to
-multiLookup :: String -> Env -> [(Env, Binder)]
-multiLookup = multiLookupInternal False
-
-multiLookupALL :: String -> Env -> [(Env, Binder)]
-multiLookupALL = multiLookupInternal True
+-- | Like 'lookupInEnv' but only returns the Binder (no Env)
+lookupBinder :: SymPath -> Env -> Maybe Binder
+lookupBinder path env = snd <$> lookupInEnv path env
 
--- TODO: Many of the local functions defined in the body of multiLookupInternal have been extracted.
--- Remove the duplication and define this in terms of the more generic/extracted functions.
-{-# ANN multiLookupInternal "HLint: ignore Eta reduce" #-}
-
--- | The advanced version of multiLookup that allows for looking into modules that are NOT imported.
--- | Perhaps this function will become unnecessary when all functions can be found through Interfaces? (even 'delete', etc.)
-multiLookupInternal :: Bool -> String -> Env -> [(Env, Binder)]
-multiLookupInternal allowLookupInAllModules name rootEnv = recursiveLookup rootEnv
-  where
-    lookupInLocalEnv :: String -> Env -> Maybe (Env, Binder)
-    lookupInLocalEnv n localEnv = case Map.lookup n (envBindings localEnv) of -- No recurse!
-      Just b -> Just (localEnv, b)
-      Nothing -> Nothing
-    importsAll :: Env -> [Env]
-    importsAll env =
-      let envs = mapMaybe (binderToEnv . snd) (Map.toList (envBindings env))
-       in envs ++ concatMap importsAll envs
-    -- Only lookup in imported modules (nonrecursively!)
-    importsNormal :: Env -> [Env]
-    importsNormal env =
-      mapMaybe (\path -> fmap getEnvFromBinder (lookupInEnv path env)) (envUseModules env)
-    importsLookup :: Env -> [(Env, Binder)]
-    importsLookup env =
-      let envs = (if allowLookupInAllModules then importsAll else importsNormal) env
-       in mapMaybe (lookupInLocalEnv name) envs
-    recursiveLookup :: Env -> [(Env, Binder)]
-    recursiveLookup env =
-      let spine = case Map.lookup name (envBindings env) of
-            Just found -> [(env, found)]
-            Nothing -> []
-          leaves = importsLookup env
-          above = case envParent env of
-            Just parent -> recursiveLookup parent
-            Nothing -> []
-       in --(trace $ "multiLookupInternal '" ++ name ++ "' " ++ show (envModuleName env) ++ ", spine: " ++ show (fmap snd spine) ++ ", leaves: " ++ show (fmap snd leaves) ++ ", above: " ++ show (fmap snd above))
-          spine ++ leaves ++ above
+-- | Like 'lookupBinder' but return the Meta for the binder, or a default empty meta.
+lookupMeta :: SymPath -> Env -> MetaData
+lookupMeta path globalEnv =
+  case lookupBinder path globalEnv of
```

Comment:
Could also write this as:

```haskell
maybe emptyMeta Meta.fromBinder . lookupBinder
```

Or with the arguments:

```haskell
maybe emptyMeta Meta.fromBinder (lookupBinder path globalEnv)
```

Reply:
The first one unfortunately didn't work, but I'll change to the second one!